### ISSUE TITLE
multi-level includes - template action can't find template in roles/<included_role>/templates/

### ISSUE THREAD

**Issue by yew011** [ansible/ansible#13818]
**Issue Type:**
- bug

**Ansible Version:**

```
ansible 2.0.0-0.9.rc4_1

ansible 2.0.0.2
```

**Ansible Configuration:**

default

**Environment:**

```
OS X Yosemite 10.10.4
```

**Summary:**

Please see the project structure below

Similar to the issue #12303 and #13671 , anther user also reports that the same issue
@nutztherookie on version 2.1.0.

**Steps To Reproduce:**

My roles layout is:

```
├── roles
│   ├── docker-service
│   │   ├── tasks
│   │   │   └── docker-service.yml
│   │   └── templates
│   │       └── docker.service.j2
│   ├── kafka
│   │   └── tasks
│   │       └── main.yml
│   └── zookeeper
│       └── tasks
│           └── main.yml
```

The **zookeeper/kafka** playbook, all include the `docker-service.yml` via:

```
- include: ../../docker-service/tasks/docker-service.yml
  vars:
    systemd_service_name: zookeeper
```

And in the docker-service playbook, it adds the systemd service file:

```
- name: adds {{ systemd_service_name }}.service files
  template:
    src: docker.service.j2
    dest: "/etc/systemd/system/{{ systemd_service_name }}.service"
    owner: root
    group: root
    mode: 0644
```

The problem is, when running, the **docker-service** playbook cannot find the
template file from `docker-service/templates/`.  And reports this error.

```
failed: [example] => (item=zookeeper) => {"failed": true, "item": "zookeeper", "msg": "IOError: [Errno 2] No such file or directory: u'/Users/alexw/tmp/projectname/docker.service.j2'"}
```

The same playbook works on Ansible 1.9.2.

**Comment by yew011** [comment]
ping?

**Comment by roberth1988** [comment]
ping?

**Comment by yew011** [comment]
I can still reproduce this issue with 2.0.0.2

Can anyone help~?!

**Comment by abadger** [comment]
We think this was probably a bug/unintended feature in 1.9.x.
- If in the docker-service yml you reference the template as ../templates/docker.service.j2 we think it should be found with both include and role because that's a relative path from the yml file.
- If the kafka and zookeeper playbooks reference docker-service as a role instead of an include then we think it should find it as the templates live in a specific location relative to a role's yml files.

In this case, however, the kafka and zookeeper playbooks reference docker-service as an include so the special directoryhandling of roles does not come into play so it seems correct that it expects the template file to live in the same directory, not in a role structure.

**Comment by yew011** [comment]
**(This reply is invalid and confusing please ignore and follow @oleksiyk  @nutztherookie reply)**

@abadger  Thx a lot for the explanation,

Good to know,

I already took the first suggestion and reference it as "../templates/docker.service.j2".

Alex Wang,

**Comment by oleksiyk** [comment]
@yew011 Does it really work for you if you reference the template in included task as "../templates/NAME" ?

I have exactly the same setup as yours, ansible 2.0.0.2 and using "../templates/NAME" simply strips one level from ansible project root dir when looking for the template, causing not found error.

**Comment by oleksiyk** [comment]
The only thing that works for me is referencing the template in included task like this:

`- template: src={{role_path}}/../included-task/templates/NAME`

**Comment by andreasnuesslein** [comment]
@oleksiyk ,

yeah - i had the same problem. @yew011 's fix wouldn't work for me either.
i will try your `{{role_path}}`-solution later. thanks!

**Comment by yew011** [comment]
@oleksiyk @nutztherookie 

Sorry my reply was very confusing,

I ended up moving the file under `zookeeper/templates/docker.service.j2`, then I could reference it.
Unfortunately, I moved away from doing this after that and forgot about this ticket.

So, let me just reopen the issue and mark my previous reply as invalid,

Thanks,
Alex Wang,

**Comment by spdionis** [comment]
I have the same issue. I confirm @oleksiyk 's solution works but would hope for a real fix.

**Comment by jdc0589** [comment]
still an issue for me: v2.0.0.2 + ubuntu 14.14. it looks like `src` is being scoped to the working directory where ansible-playbook was executed from, which seems wrong.

**Comment by jlark** [comment]
+1

**Comment by rahulinux** [comment]
Same issue with ansible 2.0.1.0

**Comment by phumpal** [comment]
Same issue with ansible 2.0.0.2.

Unfortunately, @oleksiyk's solution does not work in my case. Switching back to 1.9.4 fixes the issue.

**Comment by jan-grant** [comment]
Confirmed that this is present in 2.0.1.0.

Various attempts to locate the file with a relative path DO NOT WORK.

In my test I'm getting as far as an ssh connection - that is, the failure's on the 'remote' host:

```
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_args": {"dest": "/tmp/template.out", "src": "template.out.j2"}, "module_name": "template"}, "msg": "IOError: [Errno 2] No such file or directory: u'/path/to/top/level/playbooks/template.out.j2'"}
```

Note the file that is claimed not to exist. This from /path/to/top/level/roles/test-templates/templates/template.out.j2

**Comment by dainiax** [comment]
Same problem Ansible 2.0.1.0

**Comment by tundratim** [comment]
Any further progress on this?   Seeing this on CentOS7  with pip-installed ansible 2.0.1.0

**Comment by mixja** [comment]
This bug is now causing issues with Ansible Tower installations (at least on Centos 7), so let's hope it gets fixed soon:

```
TASK: [redis | Ensure that the Redis configuration fragments directory exists.] ***
changed: [tower01] => {"changed": true, "gid": 991, "group": "redis", "mode": "0750", "owner": "redis", "path": "/etc/redis.fragments/", "secontext": "unconfined_u:object_r:etc_t:s0", "size": 6, "state": "directory", "uid": 994}
changed: [tower02] => {"changed": true, "gid": 991, "group": "redis", "mode": "0750", "owner": "redis", "path": "/etc/redis.fragments/", "secontext": "unconfined_u:object_r:etc_t:s0", "size": 6, "state": "directory", "uid": 994}

TASK: [redis | Place the base Redis configuration file.] **********************
failed: [tower01] => {"failed": true}
msg: Source 000-base.conf.j2 not found
failed: [tower02] => {"failed": true}
msg: Source 000-base.conf.j2 not found
```

**Comment by jyrkiput** [comment]
I created a simple playbook to test different options: https://github.com/jyrkiput/ansible-role-include-task

There's two tasks (https://github.com/jyrkiput/ansible-role-include-task/blob/master/roles/role/tasks/copy.yml):

```
- name: create file
  copy: src=file dest=/tmp/file

- name: create file with relative path
  copy: src=../files/file dest=/tmp/file
```

There's also tasks using {{role_path}} (https://github.com/jyrkiput/ansible-role-include-task/blob/master/roles/role/tasks/copy_with_role_path.yml)

```
- name: create file with role path and relative
  copy: src={{role_path}}/../role/files/file dest=/tmp/file
  ignore_errors: true

- name: create file with role path
  copy: src={{role_path}}/files/file dest=/tmp/file
  ignore_errors: true
```

Including task from playbook 

```
- hosts: localhost
  tasks:
    - include: roles/role/tasks/copy.yml
      ignore_errors: true
  connection: local
```

or from another role 

```
- include: ../../role/tasks/copy.yml
```

works on 1.9.x, but fails on 2.0.x. Using {{role_path}} is problematic, as it is resolved to "roles/another_role" when included on another role, and is undefined when included from playbook. Using "{role_path}}/../role/files/file" helps a bit.

I added outputs into README (https://github.com/jyrkiput/ansible-role-include-task/blob/master/README.md)

**Comment by berlic** [comment]
I've spent some time investigating this issue.

First of all, how `include` work (as I understand it from source code):
You have a playbook – a list of tasks.
Let's assume our playbook has single play A with several tasks, one of which includes play B with some other tasks.
After you start `ansible-playbook`, ansible iterates trough tasks in play A one by one.
When there is an `include` task in the queue, ansible _copies_ all tasks from play B (our included file) into play A context.
So if there was a `template` action inside play B it will execute in the context of play A with search path relative to play A.
This is why the issue arise. But I can't say this is an unexpected behavior.

I see two ways one might want to use `include` statement:
1. We have some `common_tasks.yml` file to complete some tasks using templates/files within the current role structure. For example, we have roles `app_1` and `app_2` with specific `Dockerfile.j2` file in each app's `template` subfolder. And if include `common_tasks.yml` inside our app's roles, the call `template: src=Dockerfile.j2 ...` inside it will process each app's respective docker file template relative to role's search path.
   _This is current ansible behavior._
2. We have `roles/common_role` role (or better say role-helper that is designed to be called from other roles) that has it own templates/files. For any reason we can not describe this common_role as dependency for our roles `app_1` and `app_2`, but want to include it like this: `include: ../roles/common_role/tasks/main.yml` in the middle of our app's roles. Doing this we expect that main.yml will be executed in it's own context, using templates/files that are located under `/roles/common_role/templates` path. And here comes the issue.

So this are two different use cases for include statement.
I think both of them should be supported in ansible.
I've made a quick concept how it could be implemented: https://github.com/berlic/ansible/commit/cabc30227172de3bec628bc9637e6ad2da7f9e91

If we include file with `_raw_params`, it works as usual.
But if we call:

```
- name: include subrole
  include: role=incl_role tasks_file=main.yml
```

the hidden variable `_included_role_path` is added to each included task and template task tests for this hidden variable: if it is present, it is used as search path.
`tasks_file` may be omitted, if you want to call `main.yml`.

**Comment by stephenrlouie** [comment]
Hi all, I ran into similar issues in the reference I made in the *kubernetes/contrib* PR. New to ansible but this is what I found out today.

I have 3 levels of play books.
1. Cluster wide playbooks
2. Service orchestration playbooks
3. Service worker playbooks

For example level 1. `master-install.yml` will call `install-etcd.yml` (and other service orchestration playbooks), Then level 2. might call `etcd-download.yml` `etcd-configure.yml` `etcd-start.yml`, when I did not specify role etcd in level 2. I would see the problem described, the scope for finding templates and files was in the root directory of level 1. The work around I used (described above) was to add the paths `roles/etcd/templates/temp.j2`

What I found out was by specifying `role: etcd` in the level 2. playbook that called the roles in my `tasks` directory, resolved all of it. Granting the proper scope when looking up files and templates. 

Working example of Level 2 playbook calling a level 3
```
#deploy-etcd.yml
---
- hosts: etcd 
  sudo: yes
  #Putting the roles here makes this operate like it should, without it templates looks in the root dir where the master playbook is.
  roles:
    - etcd 
  tasks:
    - include: tasks/install.yml
    - include: tasks/configure.yml
    - include: tasks/start.yml
```

TLDR: Using the `roles` statement in addition to `include` seems to scope the "included" playbook properly, as long as the playbook being called is in the `tasks` directory of said role.

**Comment by mwisnicki** [comment]
I hit this bug with most basic setup. Just one playbook using one role with template action in main.yml. No multi-level includes. This is with ansible-2.0.2 on Linux.

**Comment by mwisnicki** [comment]
Sorry I had files out of sync. It's working with basic layout.

**Comment by bcoca** [comment]
can you test against current devel, I pushed some fixes into the search pathing recently.

**Comment by mwisnicki** [comment]
Unfortunately I can't but it's working with 2.0.2. Even with multi-level includes.

**Comment by ferrarimarco** [comment]
I tried with the latest devel and it works!

**Comment by linusgg** [comment]
@bcoca tested against 
```
ansible 2.2.0 (devel 6dc148d82c) last updated 2016/08/01 12:34:14 (GMT +1000)
  lib/ansible/modules/core: (detached HEAD 24db4de245) last updated 2016/07/27 12:20:42 (GMT +1000)
  lib/ansible/modules/extras: (detached HEAD db979dde74) last updated 2016/07/27 12:20:42 (GMT +1000)
```
```
- hosts: some_host
  roles:
    - { role: supervisor, become: yes }
```
Having this error 
`fatal: [x.x.x.x]: FAILED! => {"changed": false, "failed": true, "msg": "Unable to find 'templates/supervisord.conf.j2' in expected paths."}`
The layout inside the roles directory:
```
roles/supervisor/
├── defaults
│   └── main.yml
├── handlers
│   ├── main.yml
│   └── supervisor.yml
├── tasks
│   ├── config.yml
│   ├── install.yml
│   └── main.yml
└── templates
    └── supervisord.conf.j2
```
But it works fine with stable ansible 2.1.0.0

**Comment by linusgg** [comment]
@bcoca tested against 

```
ansible 2.2.0 (devel 6dc148d82c) last updated 2016/08/01 12:34:14 (GMT +1000)
  lib/ansible/modules/core: (detached HEAD 24db4de245) last updated 2016/07/27 12:20:42 (GMT +1000)
  lib/ansible/modules/extras: (detached HEAD db979dde74) last updated 2016/07/27 12:20:42 (GMT +1000)
```

Works good.
I want to note that while `src: templates/file.j2` and `src: file.j2` are both ok in stable 2.1.0.0 but only `src: file.j2` works in devel.
